design-proposals/klip-38-variable-substitution.md

```diff
@@ -0,0 +1,302 @@
+
+## What is not in scope
+
+* Provide support to prompt users to type a variable value if it wasn't defined (Other DBs support this)
+* Support variable substitution of environment and other non-user variables
+* Provide a set of pre-defined variables to use in the current session
+* Support variable substitution in the REST API (cannot store variables after consecutive requests)
+
+## Value/Return
+
+Users will write richer SQL scripts with customized output based on user variables.
+
+These SQL scripts may run using specific settings (topic names/format, replicas/partitions, expressions)
+allowing users to easily test the behavior in different environments (CI/CD, QA clusters, Devel/Prod environments, etc).
+
+## Public APIS
+
+* New syntax will be added to interact with user variables
+* A new CLI command and/or configuration will be added to disable/enable variable substitution
+* SQL syntax will be modified to support variable references
+
+## Design
+
+### 1. Syntax
+
+#### New syntax decision
+
+ksqlDB has a syntax to SET server and CLI properties. The use is `SET` for server properties, and `SET CLI` for CLI settings.
+
+For instance:
+```
+ksql> SET 'auto.offset.reset' = 'earliest';
+ksql> SET CLI WRAP OFF;
+```
+
+The quickest assumption is that we can use `SET` for user variables as well. But, we cannot re-use any of the above syntax because
+of conflicts with current server and CLI setting names. So a new keyword must be use.
+
+Some ideas:
+```
+SET LOCAL   var1 = 'val1';
+SET SESSION var1 = 'val1';
+SET VAR     var1 = 'val1';
+```
+
+`LOCAL` is confusing for users coming from other DBs environments. `LOCAL` is used to keep the scope on per-transaction basis. After a statement
+is executed, the variable is destroyed.
+
+`SESSION` may be also confusing. `SESSION` is used on other DBs to override system variables for that session only. System variables are like ksqlDB
+server properties.
+
+`VAR` seems a better name.
+
+However, we're now overloading the use of `SET` statements. For instance, in the future, we might want to use `SET ROLE`  to set the current role
+fo the user when interacting with an RBAC server. But, if we want to print variable information, how would we do it?
+```
+SET VAR var1;  # prints var1 name
+SET VAR;       # lists all variables
+SET;           # what to do? lists all server, cli, variables, etc?
+```
+
+Also, the use of single-quotes in `SET` server settings is tedious to type. So after removing them, we'll have two different sets:
+* `SET 'server_property' = 'server_value'`
+* `SET VAR variable = `variable_value'
+
+This looks confusing too. So we have to decide for a new syntax for substitution variables.
+
+Other DBs use different syntax to create user variables (DECLARE, DEFINE, \set, SET). For ksqlDB, I'll choose to use `DEFINE` and `UNDEFINE` which
+is what Oracle SQL*plus uses. Any user coming from this DB will immediately understand the command.  
```

Comment:
can you link to the Oracle docs here for future reference (is it https://docs.oracle.com/cd/E18283_01/server.112/e16604/ch_twelve017.htm ?)

Though if I had to choose, I think the SQL+ syntax here is a little unintuitive. MySQL: https://docs.oracle.com/cd/E17952_01/mysql-5.7-en/user-variables.html uses something that to me is more intuitive (and is similar to SqlServer)

Reply:
Oracle SQL*Plus
https://blogs.oracle.com/opal/sqlplus-101-substitution-variables

I wanted to use a similar character like Mysql. I also like the `@`, but the `StringSubstitutor` library doesn't allow me to us only the @. I have to add a suffix to the variable, so a `@var@` or `:var:` would work.

Reply:
And regarding the `SET`. I didn't want to mix the current SET syntax which uses single-quotes for keynames, and I wasn't sure what other keyword to add to make reference to user-variables. Any proposal?

```
ksql> SET 'auto.offset.reset' = 'earliest';
ksql> SET VAR var1 = 'hola';
```

Reply:
If there's precedent (Hive, Impala, Spark and Presto) I think we can stick with your proposal as users will be accustomed to this already.